src/serac/numerics/trust_region_solver.cpp

```diff
@@ -0,0 +1,481 @@
+    col_indices.push_back(i);
+  }
+
+  VecSetValues(v, local_rows, &col_indices[0], &state[0], INSERT_VALUES);
+
+  VecAssemblyBegin(v);
+  VecAssemblyEnd(v);
+
+  return v;
+}
+
+/**
+ * @brief Copy a petsc vector to an mfem::Vector
+ * @param v The petsc vector
+ * @param s The mfem vector
+ */
+void copy(const Vec& v, mfem::Vector& s)
+{
+  const int local_rows = s.Size();
+  PetscInt  iStart, iEnd;
+  VecGetOwnershipRange(v, &iStart, &iEnd);
+
+  SLIC_ERROR_IF(local_rows != iEnd - iStart,
+                "Inconsistency between local t-dof vector size and petsc start and end indices");
+
+  std::vector<int> col_indices;
+  col_indices.reserve(static_cast<size_t>(local_rows));
+  for (int i = iStart; i < iEnd; ++i) {
+    col_indices.push_back(i);
+  }
+
+  VecGetValues(v, local_rows, &col_indices[0], &s[0]);
+}
+
+/**
+ * @brief The reduced matrix in the space of {s}
+ * @param s The vector of mfem::Vector of directions
+ * @param As The vector of mfem::Vector of a global matrix A operated on directions
+ */
+Mat dot(const std::vector<const mfem::Vector*>& s, const std::vector<const mfem::Vector*>& As)
+{
+  SLIC_ERROR_IF(s.size() != As.size(),
+                "Search directions and their linear operator result must have same number of columns");
+  size_t num_cols     = s.size();
+  int    num_cols_int = static_cast<int>(num_cols);
+  Mat    sAs;
+  MatCreateSeqDense(PETSC_COMM_SELF, num_cols_int, num_cols_int, NULL, &sAs);
+  for (size_t i = 0; i < num_cols; ++i) {
+    for (size_t j = 0; j < num_cols; ++j) {
+      MatSetValue(sAs, static_cast<int>(i), static_cast<int>(j), mfem::InnerProduct(PETSC_COMM_WORLD, *s[i], *As[j]),
+                  INSERT_VALUES);
+    }
+  }
+  MatAssemblyBegin(sAs, MAT_FINAL_ASSEMBLY);
+  MatAssemblyEnd(sAs, MAT_FINAL_ASSEMBLY);
+  return sAs;
+}
+
+/**
+ * @brief The reduced vector s.T*b
+ * @param s The vector of mfem::vector of directions
+ * @param b The right hand size vector to be reduced
+ */
+Vec dot(const std::vector<const mfem::Vector*>& s, const mfem::Vector& b)
+{
+  size_t num_cols = s.size();
+  Vec    sb;
+  VecCreateSeq(PETSC_COMM_SELF, static_cast<int>(num_cols), &sb);
+  for (size_t i = 0; i < num_cols; ++i) {
+    VecSetValue(sb, static_cast<int>(i), mfem::InnerProduct(PETSC_COMM_WORLD, *s[i], b), INSERT_VALUES);
+  }
+  return sb;
+}
+
+/**
+ * @brief The qr decomposition of the state vectors
+ * @param states The vector of mfem::vectors of directions
+ */
+auto qr(const std::vector<const mfem::Vector*>& states)
+{
+  BasisVectors bvs(*states[0]);
+  BV           Q = bvs.constructBases(states);
+
+  Mat R;
+  int num_cols = static_cast<int>(states.size());
+  MatCreateSeqDense(PETSC_COMM_SELF, num_cols, num_cols, NULL, &R);
+  BVOrthogonalize(Q, R);
+
+  return std::make_pair(Q, DenseMat(R));
+}
+
+/**
+ * @brief compute the quadratic energy from small dense matrices and vectors
+ * @param A The stiffness matrix
+ * @param b The rhs vector
+ * @param x The current solution vector
```

Comment:
Does the doxygen check not throw a warning/error on missing the return parameter documentation?

Reply:
it does not seem to error.  Presumably our functions should be named such that the return is often obvious.